rust/toolchain.bzl

```diff
@@ -52,6 +52,10 @@ def build_rustc_command(ctx, toolchain, crate_name, crate_type, src, output_dir,
   extra_filename = ""
   if output_hash:
     extra_filename = "-%s" % output_hash
+    
+  # TODO consider handling 'fastbuild'
+  # opt_level and debug_info combination roughly matches Cargo's "dev" and "release" profiles
+  (opt_level, debug_info) = (0, 2) if ctx.var["COMPILATION_MODE"] == "dbg" else (3, 0)
```

Comment:
Do you have an aversion to adding support for `fastbuild` in this PR? Best way to do that is probably to pull this into a separate function, with the function returning a struct containing `opt_level` and `debug_info`. (0, 0) is probably most appropriate for `fastbuild`.

With that change, it would probably be appropriate to drop the "@TODO" from the opt-level line as well.

Reply:
@acmcarther, the problem with `(0,0)` for `fastbuild` is that `fastbuild` is the default mode.  This means we'll get a rather useless binary/lib: no debug info and no optimizations.  This is essentially why I punted on "truly" supporting `fastbuild` - I think it requires some thought, unless you think it being the default isn't really an issue and users should explicitly opt in for `opt` builds when they want performance or `dbg` if they want full debug info.

Reply:
For convenience, [here](https://docs.bazel.build/versions/master/user-manual.html#flag--compilation_mode) is the description of the 3 modes in bazel's user manual.  `debuginfo=0, opt-level=0` *is* in the spirit of `fastbuild`, but my main concern is a change of behavior for existing users that aren't explicitly specifying this option, which admittedly, seems quite C/C++ oriented (as its description even mentions).  That's also the reason I wasn't sure using `-c` was appropriate for non-C/C++ builds, even if they can mirror the debug info/opt level settings.

For my uses, I'm fine with explicitly specifying `-c opt` or `-c dbg`.  In `cargo` land, `fastbuild` is also somewhat close to `cargo check`, which is essentially "run everything up to LLMV codegen".  It doesn't produce an artifact, however, but *is* super quick :).  Although we likely can't use that for `fastbuild`.

Reply:
> my main concern is a change of behavior for existing users that aren't explicitly specifying this option, which admittedly, seems quite C/C++ oriented (as its description even mentions).

To be honest, I'm personally not overly concerned about changing the default optimization level. There have been some discussions about correctly supporting -c (opt|dbg|fastbuild) elsewhere, and I think the expectation is that they should work as they do in C++. If a user wants to have a different optimization level per compilation target, they can do so via rustc_flags. By default, I do specify an optimization level in `cargo-raze`, so I expect that this will be a no-op for many build targets.

> In cargo land, fastbuild is also somewhat close to cargo check, which is essentially "run everything up to LLMV codegen". It doesn't produce an artifact, however, but is super quick :). Although we likely can't use that for fastbuild.

I don't particularly care for `cargo check` as a basis to work forward from as far as rust_library itself is concerned but if people are interested in this functionality, I think that [Bazel Aspects](https://docs.bazel.build/versions/master/skylark/aspects.html) are a good fit. They can traverse the build tree and convert rust_library invocations into those no-output-y check-like invocations

---

All that said, take those perspectives with a grain of salt. I'm the resident Cargo/"existing rust practice" antagonist though, so /cc @mfarrugi and @damienmg for more well developed opinions about changing the default optimization level.

Reply:
I tend to default to dbg mode in my setup. I think fastbuild is not correct but is kept because in Google, binary tend to be very big so fastbuild is a better compromise.

Reply:
Also to be consistent I would just do the fastbuild do a fast build and expect people to have -c dbg on their workstation. Also probably having fastbuild on a CI is a good thing